dhis-2/dhis-support/dhis-support-test/src/main/java/org/hisp/dhis/test/setup/MetadataSetupServiceExecutor.java

```diff
@@ -0,0 +1,338 @@
+
+    private final DataElementService dataElementService;
+
+    private final IdentifiableObjectManager manager;
+
+    @Override
+    public void create( MetadataSetup setup )
+    {
+        // OBS! order matters! objects are created in such an order that
+        // referenced objects are already created
+        createEach( setup.getAttributes(), this::createAttribute );
+
+        createEach( setup.getOrganisationUnits(), this::createOrganisationUnit );
+
+        createEach( setup.getUsers(), this::createUser );
+        createEach( setup.getUserGroups(), this::createUserGroup );
+        createEach( setup.getRoles(), this::createRole );
+
+        createEach( setup.getPeriods(), this::createPeriod );
+
+        createEach( setup.getCategoryOptions(), this::createCategoryOptions );
+        createEach( setup.getCategories(), this::createCategory );
+        createEach( setup.getCategoryCombos(), this::createCategoryCombo );
+        createEach( setup.getCategoryOptionCombos(), this::createCategoryOptionCombo );
+
+        createEach( setup.getDataElements(), this::createDataElement );
+    }
+
+    private <S extends AbstractSetup<T>, T extends IdentifiableObject> void createEach(
+        Objects<S> objects, Function<S, T> creator )
+    {
+        objects.forEach( setup -> {
+            try
+            {
+                T obj = creator.apply( setup );
+                setup.setObject( obj );
+                setup.setUid( obj.getUid() );
+                Map<AttributeSetup, String> attributeValues = setup.getAttributeValues();
+                if ( !attributeValues.isEmpty() )
+                {
+                    obj.setAttributeValues( attributeValues.entrySet().stream()
+                        .map( e -> new AttributeValue( setup.getContext().getAttribute( e.getKey().getName() ),
+                            e.getValue() ) )
+                        .collect( toUnmodifiableSet() ) );
+                    manager.update( obj );
+                }
+            }
+            catch ( Exception ex )
+            {
+                log.error( "Failed to create " + setup.getClass().getSimpleName() + " " + setup.getName(), ex );
+                throw ex;
+            }
+        } );
+    }
+
+    private <S extends AbstractSetup<T>, T extends IdentifiableObject> void createEach(
+        Objects<S> objects, BiFunction<S, Objects<S>, T> creator )
+    {
+        createEach( objects, setup -> creator.apply( setup, objects ) );
+    }
+
+    private Attribute createAttribute( AttributeSetup setup )
+    {
+        Attribute obj = new Attribute( setup.getName(),
+            orElse( setup, AttributeSetup::getValueType, ValueType.TEXT ) );
+        obj.setUid( setup.getUid() );
+        obj.setUnique( setup.isUnique() );
+        obj.setMandatory( setup.isMandatory() );
+        setup.getObjectTypes().forEach( type -> obj.setAttribute( type, true ) );
+        obj.setAutoFields();
+        attributeService.addAttribute( obj );
```

Comment:
Use `IdentifiableObjectManager` where possible to reduce the number of dependencies?